desktop-exporter/app/components/waterfall-view/duration-bar.tsx

```diff
@@ -12,18 +17,35 @@ type DurationBarProps = {
 };
 
 export function DurationBar(props: DurationBarProps) {
+  const ref = useRef(null);
+  const size = useSize(ref);
+  const labelWidth = 80;
+
   let durationBarColour = useColorModeValue("cyan.800", "cyan.700");
+  let labelTextColour = useColorModeValue("blackAlpha.800", "white");
 
   let { traceStartTimeNS, traceDurationNS } = props.traceTimeAttributes;
   let spanStartTimeNs = getNsFromString(props.spanStartTimestamp);
   let spanEndTimeNs = getNsFromString(props.spanEndTimestamp);
 
-  let offsetStart = Math.floor(
+  let barOffset = Math.floor(
     ((spanStartTimeNs - traceStartTimeNS) / traceDurationNS) * 100,
   );
-  let durationBarWidth = Math.round(
+  let barWidth = Math.round(
     ((spanEndTimeNs - spanStartTimeNs) / traceDurationNS) * 100,
   );
+
+  let labelOffset;
+  if (size && size.width >= labelWidth) {
+    labelOffset = "0px";
+    labelTextColour = "white";
+  } else if (size && barOffset < 50) {
+    labelOffset = `${Math.floor(size.width)}px`;
+  } else {
+    labelOffset = `${Math.floor(-labelWidth)}px`;
+  }
```

Comment:
Some comments here on the cases would be helpful when reading

```suggestion
  if (size && size.width >= labelWidth) {
    // label is inside the bar
    labelOffset = "0px";
    labelTextColour = "white";
  } else if (size && barOffset < 50) {
    // label is ...
    labelOffset = `${Math.floor(size.width)}px`;
  } else {
    // label is ...
    labelOffset = `${Math.floor(-labelWidth)}px`;
  }
```